server/src/session.ts

```diff
@@ -131,6 +131,51 @@ export class Session {
     // project as dirty to force update the graph.
     project.markAsDirty();
     this.info(`Enabling Ivy language service for ${project.projectName}.`);
+
+    // Send diagnostics since we skipped this step when opening the file
+    // (because language service was disabled while waiting for ngcc).
+    // Technically we could just call project.refreshDiagnostics(), but we need
+    // to handle the special case where all open files are external templates.
+    const openFiles = this.getOpenFilesForNewlyLoadedProject(project);
+    this.sendPendingDiagnostics(openFiles);
+  }
+
+  /**
+   * Get the open files for a newly loaded project. If all the open files are
+   * external templates, then try to match them to the specified `project` by
+   * including one of the root files. The inclusion of a root file triggers
+   * global analysis, during which we fix up the external templates.
+   */
+  private getOpenFilesForNewlyLoadedProject(project: ts.server.Project): string[] {
```

Comment:
What do you think of splitting this into two steps/functions:

1) Get a single TS file to perform global analysis, and run `getDiagnostics` for it to do the global analysis
2) Return diagnostics for all open files

I think this would be a little easier to reason about and should be the same performance-wise.

Reply:
This is excellent suggestion! I've simplified the whole logic, albeit at a slight inefficiency -
Previously, we look at all open files, check if there is at least one TS file, and if not inject one of the root files.
With the latest change, we do (1) unconditionally, then just refresh diagnostics for the whole project.